core/trie/proof.go

```diff
@@ -133,39 +133,34 @@ func transformNode(tri *Trie, parentKey *Key, sNode StorageNode) (*Edge, *Binary
 	return edge, binary, nil
 }
 
-func traverseNodesAndAppendToPath(path *[]ProofNode, currNode *ProofNode, nodeHashes map[string]ProofNode) error {
-	whileLoop := true
-	for whileLoop {
-		switch {
-		case currNode.Binary != nil:
-			expectedLeftHash := currNode.Binary.LeftHash.String()
-			expectedRightHash := currNode.Binary.RightHash.String()
-
-			nodeLeft, leftExist := nodeHashes[expectedLeftHash]
-			nodeRight, rightExist := nodeHashes[expectedRightHash]
-			if leftExist && rightExist {
-				return errors.New("unexpected error")
-			} else if leftExist {
-				currNode = &nodeLeft
-				*path = append(*path, *currNode)
-			} else if rightExist {
-				currNode = &nodeRight
-				*path = append(*path, *currNode)
-			} else {
-				whileLoop = false
-			}
-		case currNode.Edge != nil:
-			righNode, exists := nodeHashes[currNode.Edge.Child.String()]
-			if exists {
-				currNode = &righNode
-				*path = append(*path, *currNode)
-			} else {
-				whileLoop = false
-			}
+// Begins from the root node and traverses the merged proof path
+// Until it finds a split node, adds nodes to commonPath
+// Then continues with traversing left and right paths separetaly
+// Assumes there is no circular paths and the split happens at most once
+func traverseNodes(currNode *ProofNode, path, leftPath, rightPath *[]ProofNode, nodeHashes map[felt.Felt]ProofNode) {
```

Comment:
I still think you try to do too much here :)
How about:
```go
func traverseNodes(currHash *felt.Felt, path *[]ProofNode, nodeHashes map[felt.Felt]ProofNode) {
... just populate one slice here
    if leftExist && rightExist { return }
...
```

so at the bottom of `Split...` func where you copy common to left and rigth you could:
```go
traverseNodes(rootHash, &commonPath, nodeHashes)
// this stops on split node which is the last in commonPath

splitNode := commonPath[-1] // forgive me pythonic code 😛
traverseNodes(&splitNode.Binary.LeftHash, &leftPath, nodeHashes)
traverseNodes(&splitNode.Binary.RightHash, &rightPath, nodeHashes)

// copy commons to both left & right
```

Does it make sense or I'm missing something? 🤔